### PR TITLE
C# Formatting in autoprops missing close brace

### LINKED ISSUE #7768: Formatting of automatic properties in WPF and UWP apps
Hi!

When I'm editing a WPF or UWP app in Visual Studio 2015, and I'm creating an automatic property in C#, I type something like: 
```C#
public string Whatever {get
```
but when I type the semicolon after the get, the get; moves down to the next line, regardless of my editor formatting settings in Visual Studio.
```C#
public string Whatever {
    get;
```

This does not happen in ASP.NET applications. Hardly a showstopper, but annoying and breaks my flow for a moment.

Thanks!
Scott

### PR DESCRIPTION
**Pull request by balajikris**
**TL;DR:**
With auto brace completion turned off, ensure that we keep the accessors
of a property whose close curly is missing, on the same line.

Fixes #7768, #8228.

**Long Summary:**

With Auto brace completion turned off, type `get;` at `$$` :
```C#
string Property {$$
```

it results in
```C#
string Property {
    get;
```
##### Observe:

- completion list was up
- semicolon committed the completion list and triggered the formatting.

##### Implementation Notes:

1. the semicolon first commits completion, and then invokes the formatter with the span that contains text **`{get;`**
The span to format is determined by an helper `CommonFormattingHelpers.GetFormattingSpan` which simply covers from _previous token to next token of a given span_. Formatting this span puts `get` on the next line.
 
2. the semicolon then triggers formatting, with the span that encompasses the text **`get;`** Notice that the formatter now arrives at a different span to format than the above. this is because it uses `FormattingRangeHelper.FindAppropriateRange` which has smarts about different scenarios and notably has adjustments for including or excluding the leading open brace from formatting. So essentially, this would not move the accessor to the new line, but for this case though, the newline has already been put in by step 1.

_So, we have a few ways to fix this :_

1. Do not invoke formatting from completion list
2. Fix the helper (`GetFormattingSpan`) to determine spans to include knowledge about when to include/exclude the leading open brace. 
3. Fix the general case for auto properties, that when their close brace is missing and the syntax node has an error, do not force the accessor onto a new line. This should work even when invoked from format document or format selection.

_This PR implements approach 3, and creates 2 future work items to :_

1. #8337: Investigate and understand why we need to invoke the formatter from completion list, especially in this case when semicolon is a trigger character for formatting as well.
2. #8338: Possibly consolidate the helper methods in `FormattingRangeHelper` and `CommonFormattingHelpers` classes that are used to determine the span/range to format

### CODE REVIEW

**Comment by Pilchie** [comment]
@CyrusNajmabadi has made some changes to formatting in this kind of case and should take a look

**Comment by balajikris** [comment]
@Pilchie : Yes, I spoke with @CyrusNajmabadi about this.
tagging for review: @dotnet/roslyn-ide @CyrusNajmabadi @heejaechang @basoundr

**Comment by basoundr** [comment]
The right fix here is to make sure completion list gives the correct span to format to formatter instead of changing the rule.

The proposed fix to create an exception for incomplete property declaration is just handling one case of the main problem(incorrect span to format)

The same problem will happen in below case as well.

```C#
class Program
{
    public void M() { return;$$

    static void Main(string[] args)
    {
    }
}
```

**Comment by balajikris** [comment]
@basoundr Thanks for the feedback. I'll try to expand a bit more on that.

1. The span to format is being determined by the formatting engine in both cases (triggered from completion, triggered by formatter). the difference arises because these are different APIs within the formatting service and they use different helpers to determine the span to format. There could be good reasons to do so, but I am not sure what they are yet. The api that completion hits, is also hit by VB Line commit and a couple more features (forget their names but i can look it up). So, one of my concerns is making the change there has high impact given the time of the product cycle we're in.

2. Since you say the fix should be the span to format, I'd like to note that format selection or format document will still run into the problem. Although, I concede that invoking format selection or document mid way through writing a statement doesn't feel like a compelling scenario or one that blocks typing fluency to me.

3. your example seems a bit contrived to me (a single line void returning method), is there a broader real case that you are concerned about ?

4. Finally, If you feel the fix is hacky, then, certainly, I'm all for not doing it this way.

Thoughts?

**Comment by basoundr** [comment]
_On Point 1_. It is perfectly reasonable to introduce an formatting helper which addresses the scenarios of completion list. 

_On Point 2_. Just like you mentioned, if the user is going to explicitly ask for formatting then we have to format 

_On Point 3_. This is a perfectly valid scenario where the user might have an overriden method for a base method where he/she decides to do nothing.
```C#
class TestProgramDontDoAnything
{
    public override void UpdateRecords() { return; }

    public void Main()
    {
    }
}
```

Yes, I do feel this fix is hacky and if there are concerns that the right fix might affect the product cycle then I would advice the issue to be moved over to the next release. Not because the fix is hacky but also this is the not the right fix and the problem still persists after the fix

**Comment by balajikris** [comment]
okay, let me take another shot at fixing this. I'm going to try determining the right span to format from completion.

/cc: @CyrusNajmabadi since we discussed this earlier.

**Comment by CyrusNajmabadi** [comment]
> The right fix here is to make sure completion list gives the correct span to format to formatter instead of changing the rule.

I'm not sure i agree.  It seems very odd to me that you'd complete and get different formatting than if you just said "format".  Now we have two different sets of behaviors depending on if a completion list was up or not when you just semicolon.

I think i'd far prefer it if hte rule was just consistent regardless of what span was being formatted.  Why would we want all this "extra" logic in the completion case that we now have to internalize and understand?

Thanks!

File: src/Workspaces/CSharp/Portable/Extensions/SyntaxTokenExtensions.cs [hunk 1]
```diff
@@ -354,5 +354,17 @@ public static bool IsOpenBraceOfAccessorList(this SyntaxToken token)
         {
             return token.IsKind(SyntaxKind.OpenBraceToken) && token.Parent.IsKind(SyntaxKind.AccessorList);
         }
+
+        public static bool IsOpenBraceOfPropertyAccessorList(this SyntaxToken token)
+        {
+            return token.IsOpenBraceOfAccessorList() && (token.Parent?.IsParentKind(SyntaxKind.PropertyDeclaration) ?? false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/SyntaxTokenExtensions.cs]: no needd for ?.  IsParentKind is resilient to null.

**Comment by balajikris** [comment]
@CyrusNajmabadi and @basoundr 

I've tried the other approach at #8425. Notice a slight difference in output between the two approaches 

With the new approach (#8425) there won't be a newline before the `get` accessor or a `space` before it either. the span between `{` and `get` won't be formatted at all for the semicolon. The required spacing would be put in when the user completes the declaration with a close brace `}`. 
```C#
int P {get;
```
Whereas with this PR (#8336), we'd have gotten the desired space after typing the semicolon. 
```C#
int P { get;
```

However, the difference isn't bad at all, because, (a) we're getting exactly the same formatting we'd have gotten if completion was not up. (b) closing the construct with close brace will put in the space soon anyway.

**Comment by CyrusNajmabadi** [comment]
> (a) we're getting exactly the same formatting we'd have gotten if completion was not up

Then this is good :)

**Comment by balajikris** [comment]
corrected my text above a bit to not be ambiguous - "However, _the difference in case of the new approach_ isn't bad at all,"

**Comment by basoundr** [comment]
There are a couple of reasons, I said it is not the right fix.

* There is a separation of responsibilities when it comes to formatting. The _rules_ produce operations that needs to be applied on the code and the _engine_ chooses to apply the operations or not. Now, this particular change might give the behavior that we are hoping for, but the rules are doing more than what they are supposed to do. It is doing the engine's job. Also, consider the example I mentioned. This would mean that, we need to create a bunch of exceptions for all possible broken code cases and that will clutter the code.

* I don't think I can agree with asking for the same behavior for 2 different selected spans, given that the span includes broken code because there is a lack of information for formatting to be consistent. Given that Completion (or any other feature for that matter) is more aware of the state of the code, Completion should make sure to provide the appropriate information to the formatting engine. The expectation of Formatting Engine to have a heuristics to expect that the user is trying to type everything in the same line is going to make Formatting slower and introduce more heuristics, which is both not appropriate and will make formatting engine slow

I feel #8425 is a reasonable approach given what I have mentioned above

**Comment by balajikris** [comment]
Closing this in favor of #8425.

**Comment by CyrusNajmabadi** [comment]
> The rules produce operations that needs to be applied on the code and the engine chooses to apply the operations or not.

If htat's the case, then i'm fine with having the *engine* to decide not to run things here because the code is broken.  

> Given that Completion (or any other feature for that matter) is more aware of the state of the code, Completion should make sure to provide the appropriate information to the formatting engine. 

I disagree.  It's not even clear to me why completion is invoking formatting in the first place.  Why is completion even involved?  And even if we think completion should be involved, why would it need to generate a different span ever than what formatting would normally produce when deciding what to format here.

Here's a critical thing that's bothering me:

1.  The typing *without* completion case works fine.  We produce an appropriate span and the rules/engine properly decide not to format this code.
2. With completion things go off the rails.  Why?  Why is completion differing here?  completion doesn't have any more "awareness" of the "State of the code".  

It seems very weird to me that the core engine/rules/etc. is working properly and a feature is injecting code that makes that stop happening.  I'd far prefer:
1. completion actually invoke formatting.
2. if completion has to invoke formatting, that it uses the exact same helper to get the span to format as normal typing does.

Having completion be entirely different just means it becomes vastly more difficult to figure out how formatting is supposed to work.

**Comment by balajikris** [comment]
Addressing some questions above:

> Why is completion differing here?

bk>> It is using a different helper. See my comment above in the original PR request. Why there are many helpers, I don't know but i filed a work item to get it cleaned up.
> #8338: Possibly consolidate the helper methods in FormattingRangeHelper and CommonFormattingHelpers classes that are used to determine the span/range to format

--
>1.completion actually invoke formatting.

bk>> I think you mean "Completion actually **not** invoke formatting". If so, that is deferred to the future workitem below.

> #8337: Investigate and understand why we need to invoke the formatter from completion list, especially in this case when semicolon is a trigger character for formatting as well.

--
> 2.if completion has to invoke formatting, that it uses the exact same helper to get the span to format as normal typing does.

bk>> This is what I have done with #8425.

**Comment by CyrusNajmabadi** [comment]
Great.  I'll check out #8425.  It sounds like it's the right short term fix.  Thanks!

**Comment by heejaechang** [comment]
I agree with @basoundr formatting engine doing right thing by default in all cases never worked. different features that require slightly different behavior should take care of that themselves.

formatting engine should have default behavior (for broken code) and default rules toward to format entire document case and any feature that wants slightly different behavior (code generator wanting force 1 blank line between methods or auto formatting on ";" doesn't want to break lines between broken codes or smart indenter on "," want to align to previous "," and etc) should add those rule in that feature.

this isolation actually should reduce dependency between features and let them change its behavior without worrying about affecting others.

**Comment by heejaechang** [comment]
for this particular case, I think we should either we do what @basoundr suggested - giving right span - or add completing set specific rule to formatting engine such as this code in completion set code.
